<a href="https://colab.research.google.com/github/htapiagroup/sistemas-distribuidos-evaluacion-intersemestral-cflores23/blob/master/Sist_Dist2_Examen1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Primer examen Parcial

Alumno: Camilo Flores Rodríguez

Trabajé solo

#Importar datos
      1. Adquirir datos
      Para adquirir datos descargo con la instrucción del S.O. !wget
      
      He utilizado mi Google drive por lo que debo motar la unidad

In [0]:
!wget https://archive.org/download/nycTaxiTripData2013/trip_data.7z

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')


Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


Observo donde me encuentro ubicado:

In [3]:
ls

gdrive/  sample_data/


In [3]:
ls


gdrive/  sample_data/


Puedo observar que sí se descargó en mi unidad drive los archivos que necesito para el ejercicio.

Como ya he montado mi unidad de Google drive me cambio a la carpeta Colab Notebooks

In [4]:
cd gdrive/My\ Drive/'Colab Notebooks'

/content/gdrive/My Drive/Colab Notebooks


In [0]:
Para poder responder la pregunta 3 necesito utilizar la liberría Pandas por lo que la imoporto de una vez

In [0]:
import pandas as pd, time

3.  Una vez descargado el archivo pueden descomprimirlo usando la siguiente instruccion

Continúo con descomprimir un archivo

In [5]:
!7z e trip_data.7z -o/content/gdrive/My\ Drive/Colab\ Notebooks/ExamData trip_data_1.csv


7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,2 CPUs Intel(R) Xeon(R) CPU @ 2.30GHz (306F0),ASM,AES-NI)

Scanning the drive for archives:
  0M Scan         1 file, 4102781969 bytes (3913 MiB)

Extracting archive: trip_data.7z
--
Path = trip_data.7z
Type = 7z
Physical Size = 4102781969
Headers Size = 433
Method = LZMA:24
Solid = -
Blocks = 12

  0%    
Would you like to replace the existing file:
  Path:     /content/gdrive/My Drive/Colab Notebooks/ExamData/trip_data_1.csv
  Size:     2459600863 bytes (2346 MiB)
  Modified: 2014-05-12 22:02:53
with the file from archive:
  Path:     trip_data_1.csv
  Size:     2459600863 bytes (2346 MiB)
  Modified: 2014-05-12 22:02:53
? (Y)es / (N)o / (A)lways / (S)kip all / A(u)to rename all / (Q)uit? Y

  0% - trip_data_1.csv                        1% - trip_data_1.csv

Busco donde se descomprimió el archivo

In [6]:
ls

'Copia de Copia de SisDist_S1_L1.ipynb'                Examen_Final.ipynb
'Copia de Copy of MCA_SistDist_Ex1'                    Sesion2_Sist_dist2.ipynb
'Copia de Hola, Colaboratory'                          SisDist2_Ex1.ipynb
'Copia de Hola, Colaboratory (1)'                     'SisDist2_S2 (1).ipynb'
'Copia de Hola, Colaboratory (2)'                      SisDist2_S2.ipynb
'Copia de Sesion2_Sist_dist2.ipynb'                    SisDist2_sesion2.ipynb
'Copia de SisDist_S1_L1.ipynb'                         SisDist_S1_L1.ipynb
'Copia de SisDist_S5.ipynb'                            SisDist_S5.ipynb
'Copia de SistDist_S4 (1).ipynb'                       SistDist2/
'Copia de SistDist_S4.ipynb'                           SistDist2_Exam1.ipynb
'Copia de SistDist_S6'                                 Sist_Dist2_Examen1.ipynb
'Copia de Te damos la bienvenida a Colaboratory'       SistDist_S4.ipynb
'Copia de Te damos la bienvenida a Colaboratory (1)'   SistDist_S6.ipynb
'Copia de Te damos la bienv

El archivo se encuentra en la carpeta que cree ExamData, por lo tanto me cambio de directorio

In [7]:
cd ExamData

/content/gdrive/My Drive/Colab Notebooks/ExamData


In [8]:
ls

trip_data_1.csv  Untitled0.ipynb


Puedo observar de la celda anterior que el archivo trip_data_1.csv si se encuentra en la carpeta

Leo el archivo con pandas

In [0]:
df = pd.read_csv('trip_data_1.csv')

Verifico si lo leyó bien, para ello visualizo los primeros datos con los nombres de las columnas

In [10]:
df.head()

medallion  ... dropoff_latitude
0  89D227B655E5C82AECF13C3F540D4CF4  ...        40.751171
1  0BD7C8F5BA12B88E0B67BED28BEA73D8  ...        40.750660
2  0BD7C8F5BA12B88E0B67BED28BEA73D8  ...        40.726002
3  DFD2202EE08F7A8DC9A57B02ACB81FE2  ...        40.759388
4  DFD2202EE08F7A8DC9A57B02ACB81FE2  ...        40.747868

[5 rows x 14 columns]

Respondo a las preguntas 

3. Responder las siguientes preguntas:

    1.  Cuantas columnas contiene cada archivo de datos descomprimido
    2.  Cuantos renglones tiene cada archivo
    
    Para ello utilizo shape de pandas y me da la respuesta que busco

In [11]:
df.shape

(14776615, 14)

In [0]:
boleana = pd.isnull(df)
print (boleana)

De acuerdo a la consulta anterior tenemos 14 columnas y 14776615 líneas o registros

4. Identificar si existen renglones con errores en los datos, por ejemplo si hay columnas de mas (o de menos), si hay campos vacios, etc.

    1. indicar claramente cuantos son los renglones con errores y a que archivos corresponden

In [30]:
nulos=len(df[pd.isnull(df.medallion)])
print(nulos)

0


Para responder si hay o no valores nulos esta última instrucción nos sirve, solo que debemos iterarla para que lea todas las columnas.

Voy a crear una lista con los nombres de las columnas del archivo y despues recorreré la lista con un for, para determinar si tienen o no valores nulos

#Columnas con valores nulos:

In [28]:
nombres_cols = df.columns.tolist()

for column in nombres_cols:
  print("Valores nulos en la columna <{0}>: {1}".format(column, df[column].isnull().sum()))


Valores nulos en la columna <medallion>: 0
Valores nulos en la columna <hack_license>: 0
Valores nulos en la columna <vendor_id>: 0
Valores nulos en la columna <rate_code>: 0
Valores nulos en la columna <store_and_fwd_flag>: 7326207
Valores nulos en la columna <pickup_datetime>: 0
Valores nulos en la columna <dropoff_datetime>: 0
Valores nulos en la columna <passenger_count>: 0
Valores nulos en la columna <trip_time_in_secs>: 0
Valores nulos en la columna <trip_distance>: 0
Valores nulos en la columna <pickup_longitude>: 0
Valores nulos en la columna <pickup_latitude>: 0
Valores nulos en la columna <dropoff_longitude>: 86
Valores nulos en la columna <dropoff_latitude>: 86


Podemos observar que nuestro archivo si tiene valores nulos

     2. crear un nuevo conjunto de datos con los errores eliminados

In [0]:
df_sinNulos = df.dropna(how='any')

Ahora volvamos a correr la instrucción de la celda 28 para verificar si se eliminaron los nulos

In [32]:
nombres_cols = df_sinNulos.columns.tolist()

for column in nombres_cols:
  print("Valores nulos en la columna <{0}>: {1}".format(column, df_sinNulos[column].isnull().sum()))

Valores nulos en la columna <medallion>: 0
Valores nulos en la columna <hack_license>: 0
Valores nulos en la columna <vendor_id>: 0
Valores nulos en la columna <rate_code>: 0
Valores nulos en la columna <store_and_fwd_flag>: 0
Valores nulos en la columna <pickup_datetime>: 0
Valores nulos en la columna <dropoff_datetime>: 0
Valores nulos en la columna <passenger_count>: 0
Valores nulos en la columna <trip_time_in_secs>: 0
Valores nulos en la columna <trip_distance>: 0
Valores nulos en la columna <pickup_longitude>: 0
Valores nulos en la columna <pickup_latitude>: 0
Valores nulos en la columna <dropoff_longitude>: 0
Valores nulos en la columna <dropoff_latitude>: 0


Como podemos observar en el resultado anterior, hemos creado un nuevo conjunto llamado df_sinNulos sin los errores 

Para hacer el cómputo más rápido y eficiente dividiré el archivo en trozos.

Debido a que el archvo contiene 14776615 lineas lo dividiré primero en tamaños de 500,000 y probaremos si este tamaño es eficiente.

In [0]:
archivo = 'trip_data_1.csv'

#5. Crear un DataFrame usando la libreria pandas y responder lo siguiente:

    1. Indicar el tiempo en segundos, que tarda la libreria pandas en leer un archivo a un DataFrame
    2. Cual es la huella de memoria del proceso usado para generar el objeto DataFrame
    3. Cuanta memoria del sistema se usa para crear el objeto anterior
    4. Indicar el tiempo que tarda pandas en obtener el promedio de la distancia de viaje (trip_distance)

El dataframe ya se ha creado arriba para resolver las preguntas anteriores. Sin embargo, para fines de responder las del punto 5 lo volveremos a crear utilizando los mismos datos

In [0]:
t_start = time.time()
df = pd.read_csv('trip_data_1.csv')
t_end = time.time()
print('Tiempo de lectura: {} s'.format(t_end-t_start)) # time [s]

Definimos una función que nos permita medir la huella de memoria

In [0]:
import psutil, os

def huella_memoria():
  '''Regresa la huella de memoria usada por un proceso de Python'''
  mem = psutil.Process(os.getpid()).memory_info().rss
  return (mem / 1024**2)

In [0]:
antes = huella_memoria()

In [0]:
n = (1024**2)//8

In [0]:
df = pd.read_csv('trip_data_1.csv')

In [0]:
despues = huella_memoria()

Respondemos pregunta 2

In [14]:
print('Memoria antes: {} MB'.format(antes))

Memoria antes: 143.6328125 MB


In [15]:
print('Memoria despues: {} MB'.format(despues))

Memoria despues: 5107.66796875 MB


Para responder la pregunta 3 hacemos una resta para saber la diferencia entre la memoria anterior y la actual, así sabremos cuanta memoria ocupa el objeto.

In [16]:
mem_objeto = despues-antes

print('La cantidad de memoria que ocupa el dataframe es de: {} MB'.format(mem_objeto))

La cantidad de memoria que ocupa el dataframe es de: 4964.03515625 MB


In [20]:
promedio=df['trip_distance'].mean()
print(promedio)

2.770975670679966


In [22]:
for trozo in pd.read_csv(archivo, chunksize=500000):
  print('type: %s shape %s' % (type(trozo), trozo.shape))

type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (5000